### ISSUE TITLE
Proposal: Metadata to indicate the value of preserving output, e.g. tag discardable / low-value output for caching or post-mortem

### ISSUE THREAD

**Issue by whlavina** [common-workflow-language/common-workflow-language#856]
In practice, some outputs of a workflow may be very voluminous but of low value to persist beyond immediate execution, and it would be useful to be able to tag such cases, say, with a hint for workflow runners that indicates the value of preserving certain outputs:
* One use case is for caching, if such outputs are easily reproduced, in which case storing the intermediate results is sub-optimal.
* Another use case is for result validation and postmortem, when keeping the intermediate, as with ```cwltool --leave-outputs```, where some of these intermediate require excessive space, both during runtime and possible for future preservation.

For the first use case, a smart runner might evaluate the runtime cost of a step versus the output size, and automatically decide which intermediates to prune from the cache.

For the second use case, an ad-hoc postprocessing script could prune outputs left for post-mortem, but then the question remains how to decide which ones to prune, without some metadata. It would be convenient to have a way to flag outputs which are essential for analyis vs outputs which are limited value beyond workflow execution vs outputs without assertions of value.

A somewhat related proposal is #583, although that proposal dealt with reproducibility issues rather than pure performance/cost optimization.

**Comment by kaushik-work** [comment]
@whlavina possibly `workreuse` has made it to V1.1 (https://www.commonwl.org/v1.1.0-dev1/CommandLineTool.html#WorkReuse).

**Comment by whlavina** [comment]
```workreuse```, from #583 as I linked, is somewhat related, but I'm not sure it fully addresses the use case here. This one is more about having a debug mode, for scientific review of results as well as postmortem of failures, where certain intermediates are preserved and others discarded, as opposed to production mode where only the final outputs are really of interest, with intermediates being purely a caching optimization.

In SnakeMake, there is a feature for identifying [Protected and Temporary Files](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#protected-and-temporary-files) via ```protected()``` and ```temp()``` decorators.

In GNU Make, there is the feature of ```.PRECIOUS``` and ```.INTERMEDIATE``` special targets, to flag important versus discardable outputs.

The distinction is those 2 attributes indicate not just that data might get preserved, but moreso, that it _should_ be preserved somewhere for debugging (in the former case) or _should_ be discarded as soon as possible (in the latter) due to resource constraints on disk space. Automated heuristics can probably address the majority of cases for deciding which intermediates are useful to preserve for purposes of optimizing space vs compute during retry scenarios, but manual overrides that specifically indicate outputs known to be highly valuable for diagnostics and/or highly wasteful of resources, may be a useful addition.

As an alternative for marking intermediates for preservation, one may of course emit the intermediate output as final output of the workflow, but then they will always be emitted, as opposed to doing so only during a "debug" mode.

If CWL supported conditional nodes (conditionals and iteration being a known feature request for the future), then it would be possible to disable passing intermediate outputs as final output of the workflow; currently, another hackish workaround could be to have steps that always execute but conditionally emit empty output.

In WDL, there's ```call-caching``` which seems homologous to ```workreuse```, but I don't see anything similar for the debugging use case I am proposing.

**Comment by geoffjentry** [comment]
> In WDL, there's call-caching which seems homologous to workreuse, but I don't see anything similar for the debugging use case I am proposing.

To be clear, WDL does not have call caching or anything similar in the spec. **Cromwell** has this feature and it applies to both WDL and CWL

**Comment by tetron** [comment]
In the CWL sense, outputs are objects (Files and Directories) that will be passed to a downstream step or as final output.  In cwltool specifically, the default behavior (without call caching) is to retain outputs until the end of the workflow, move all the bound outputs to the final output directory, and delete the intermediate directories.  

Currently cwltool doesn't distinguish output files that are bound to output parameters, but that would be an optimization (if there's a large file in the output directory that isn't bound to an output parameter, it can be deleted.)  However, If your workflow produces temporary files that should always be cleaned up at the end of execution, you should probably write them to `$(runtime.tmpdir)` which is deleted at the end of the run (unless `--leave-tmpdir` is specified).

Cwltool also doesn't distinguish output files that are consumed by a downstream step, but not propagated any further.  Another optimization would be to delete those files after they are no longer needed by any pending steps.

If you want to capture outputs conditionally, I you should already be able do something like:

```
inputs:
  debug: boolean
outputs:
  diagnostics:
    type: File?
    outputBinding:
      glob: $(inputs.debug ? "debug.log" : null)
```

**Comment by whlavina** [comment]
The pragmatic constraint is that in non-trivial workflows, there will be large intermediates that need to be passed between 2 tasks, and that implies they must be outputs (not in ```tmpdir```, as I understand), and keeping all intermediates (for caching, diagnostics, or post-mortem) requires significant space.

**Comment by tetron** [comment]
@whlavina Yes, so that requires improving the workflow runner so that it can identify when files are no longer needed.  However I think that's an implementation feature rather than something that needs to be specified by CWL itself (we should already be able to identify which files are precious, they are the ones that contribute to the final output).

**Comment by whlavina** [comment]
Clarifying: A runner may be able to automatically make some judgement calls based on space/time/compute calculations, but the feature I am requesting here regarding the CWL specification, is formalizing a standard hint to inform the runner, via manual (human-curated) annotation, that which is known to be important or unimportant.

If you believe such hints are not useful, feel free to close this issue as something you won't do, but my opinion is that it is useful and I tried to show how at least some other tools have support for the concept.

**Comment by mr-c** [comment]
@whlavina I think such a hint is interesting.

To make this happen I suggest the following:

Working with a specific CWL aware workflow runner, make a namespaced hint for your proposal and implement it in that runner. Let us know about the name of the extension and share any lessons learned from implementing it. If there is continued interest from the community and it also gets implemented in the CWL reference runner `cwltool` then it can be a candidate for inclusion as an optional feature for the next release of the CWL standards.

Examples of other extensions: 

https://doc.arvados.org/user/cwl/cwl-extensions.html
https://github.com/rabix/bunny/wiki/CWL-draft-2-extensions

**Comment by whlavina** [comment]
Thanks for the feedback and guidance for implementation. I may not have time to attempt coding a proof of concept at this time, but I may give it a shot in the near future, or maybe someone else may like the idea and beat us to it. I'll let you decide on the resolution status of this ticket (whether you keep it open with the labels, or if you have a status for tickets that are temporarily stalled).
